tests/test_calculate_imputation_factors.py

```diff
@@ -87,98 +98,136 @@ def test_wrangler(self):
-                print(mock_json.dumps.called)
-                print(mock_json.dumps.call_args)
-                assert mock_json.dumps.call_args[0][0]['success']
-                assert response
+        assert_frame_equal(
+            actual_outcome_dataframe.astype(str), expected_output_dataframe.astype(str)
+        )
+
+    def test_wrangler_exception_handling(self):
+        """
+        testing the exception handler works within the wrangler.
+        :param self:
+        :return: mock response
+        """
+        response = calculate_imputation_factors_wrangler.lambda_handler(None, None)
+        assert not response["success"]
+
+    def test_method_exception_handling(self):
+        """
+        testing the exception handler within the method.
+        :param self:
+        :return: mock response
+        """
+
+        json_data_content = (
+            '[{"movement_Q601_asphalting_sand":0.0},'
+            '{"movement_Q601_asphalting_sand":0.857614899}]'
+        )
+
+        response = calculate_imputation_factors_method.lambda_handler(
+            json_data_content, None
+        )
+        assert not response["success"]
+
+    def test_get_traceback(self):
+        traceback = calculate_imputation_factors_wrangler._get_traceback(
+            Exception("Mike")
+        )
+        assert traceback == "Exception: Mike\n"
+
+    @mock_sqs
+    def test_marshmallow_raises_wrangler_exception(self):
+        sqs = boto3.resource("sqs", region_name="eu-west-2")
+        sqs.create_queue(QueueName="test_queue")
+        queue_url = sqs.get_queue_by_name(QueueName="test_queue").url
+        # Method
+        with mock.patch.dict(
+            calculate_imputation_factors_wrangler.os.environ, {"queue_url": queue_url}
+        ):
+            calculate_imputation_factors_wrangler.os.environ.pop("method_name")
+            out = calculate_imputation_factors_wrangler.lambda_handler(
+                {"RuntimeVariables": {"checkpoint": 666}}, None
+            )
+            self.assertRaises(ValueError)
+            print(out)
+            assert out["error"].__contains__(
+                """ValueError: Error validating environment params:"""
+            )
+
+    @mock_sqs
+    def test_marshmallow_raises_method_exception(self):
+        sqs = boto3.resource("sqs", region_name="eu-west-2")
+        sqs.create_queue(QueueName="test_queue")
+        queue_url = sqs.get_queue_by_name(QueueName="test_queue").url
+        # Method
+        with mock.patch.dict(
+            calculate_imputation_factors_method.os.environ, {"queue_url": queue_url}
+        ):
+            calculate_imputation_factors_method.os.environ.pop("questions")
+            out = calculate_imputation_factors_method.lambda_handler(
+                {"RuntimeVariables": {"checkpoint": 666}}, None
+            )
+            self.assertRaises(ValueError)
+
+            assert out["error"].__contains__(
```

Comment:
Using 'magic methods' like this is a code smell...these are bound to python descriptors so you don't need to call it directly.

Your test should be `assert "ValueError: Error validating environment params:" in out["error"]`, underneath python converts `in` to calling the `__contains__` method so it's essentially the same.